File: src/demos/ps/oo/Enemy.java
package demos.ps.oo;

import demos.ps.oo.PSLibSound.PS1Sound;
import domain.CHR;

//Following the Builder pattern for clarity

public class Enemy {

	enum Type{NONE, PALMA, MOTAVIA, DEZORIS, SPECIAL, UNDEAD};
	
	enum HasItem{NONE, COLA, DIMATE, FLASH};
	enum CanTalk{YES, NO};
	enum CanChat{YES, NO};
	enum CanRope{YES, NO};
	enum CanProt{YES, NO};
	enum HasWing{YES, NO};
	enum FireRes{YES, NO};
	enum Special{NONE, FIRE, THUNDER, THUNDER2, ROPE, HELP, PETRIFY, CURE, MP_DRAIN, DOUBLE_ATTACK};
	enum Mental{LOWEST, LOWER, NORMAL, HIGHER};
		
	private String name;
	int hp=0, atk=0, def=0, exp=0, mst=0, num=1, run=0, trap=0;
	int specialShiftX = 0, specialShiftY = 0;
	 
	// Optional
	CanTalk talk=CanTalk.NO;
	CanChat chat=CanChat.NO;
	CanRope rope=CanRope.YES;
	CanProt prot=CanProt.YES;
	HasWing wing=HasWing.NO;
	FireRes fire=FireRes.NO;
	HasItem item=HasItem.NONE;
	Special special = Special.NONE;
	Mental mental = Mental.NORMAL;
	
	PS1Sound attackSound;
	Type type;
	 
	String strAnimCHR;
	CHR animCHR;
	int vertical;
	int contact;
 
	public Enemy(String name) {
		this.name = name;
	}
	
	public Enemy hp(int hp) {
		this.hp = hp;
		return this;
	}
	public Enemy atk(int atk) {
		this.atk = atk;
		return this;
	}
	public Enemy def(int def) {
		this.def = def;
		return this;
	}
	public Enemy exp(int exp) {
		this.exp = exp;
		return this;
	}
	public Enemy mst(int mst) {
		this.mst = mst;
		return this;
	}
	public Enemy num(int num) {
		this.num = num;
		return this;
	}
	public Enemy run(int run) {
		this.run = run;
		return this;
	}	
	public Enemy trap(int trap) {
		this.trap = trap;
		return this;
	}
	
	// Optional (has proper default values)
	public Enemy item(HasItem item) {
		this.item = item;
		return this;
	}	
	public Enemy talk(CanTalk talk) {
		this.talk = talk;
		return this;
	}	
	public Enemy chat(CanChat chat) {
		this.chat = chat;
		return this;
	}	
	public Enemy rope(CanRope rope) {
		this.rope = rope;
		return this;
	}	
	public Enemy prot(CanProt prot) {
		this.prot = prot;
		return this;
	}	
	public Enemy wing(HasWing wing) {
		this.wing = wing;
		return this;
	}
	public Enemy fire(FireRes fire) {
		this.fire = fire;
		return this;
	}	
	public Enemy special(Special special) {
		this.special = special;
		return this;
	}

	public Enemy mental(Mental mental) {
		this.mental = mental;
		return this;
	}
	
	public Enemy spcpoint(int i, int j) {
		this.specialShiftX = i;
		this.specialShiftY = j;
		return this;
	}

	public Enemy sound(PS1Sound sound) {
		this.attackSound = sound;
		return this;
	}
	
	public PS1Sound getSound() {
		return this.attackSound;
	}
	
	public int getMaxHp() {
		return hp;
	}
	public int getMaxNum() {
		return num;
	}
	
	public Enemy anim(String strAnim) {
		this.strAnimCHR = strAnim;
		return this;
	}	

	public Enemy vertical(int i) {
		this.vertical = i;
		return this;
	}
	public Enemy contact(int i) {
		this.contact = i;
		return this;
	}
	public Enemy type(Type type) {
		this.type = type;
		return this;
	}
	

	// Lazy load CHR
	public CHR getChr() {
		if(this.animCHR == null) {
			this.animCHR = CHR.loadChr(this.strAnimCHR);			
		}
		return this.animCHR;
	}

	public int getMental() {
		switch(mental) {
		case HIGHER:
			return this.getMaxHp() * 2;
		case LOWER:
			return this.getMaxHp() / 2;
		case LOWEST:
			return this.getMaxHp() / 4;
		default:
			return this.getMaxHp();
		}
	}
	
	public String getName() {
		return PSGame.getString(this.name);
	}

}

File: src/demos/ps/game/Enemy.ts
/**
 * Enemy - Battle Enemy System
 * Defines enemy entities with builder pattern
 */

import { PS1Sound } from './PSLibSound';
import { PSGame } from '../PSGame';
import { CHR } from '../../../domain/CHR';

export enum EnemyType {
  NONE,
  PALMA,
  MOTAVIA,
  DEZORIS,
  SPECIAL,
  UNDEAD
}

export enum HasItem {
  NONE,
  COLA,
  DIMATE,
  FLASH
}

export enum CanTalk {
  YES,
  NO
}

export enum CanChat {
  YES,
  NO
}

export enum CanRope {
  YES,
  NO
}

export enum CanProt {
  YES,
  NO
}

export enum HasWing {
  YES,
  NO
}

export enum FireRes {
  YES,
  NO
}

export enum Special {
  NONE,
  FIRE,
  THUNDER,
  THUNDER2,
  ROPE,
  HELP,
  PETRIFY,
  CURE,
  MP_DRAIN,
  DOUBLE_ATTACK
}

export enum Mental {
  LOWEST,
  LOWER,
  NORMAL,
  HIGHER
}

export class Enemy {
  private name: string;
  public hp: number = 0;
  public atk: number = 0;
  public def: number = 0;
  public exp: number = 0;
  public mst: number = 0;
  public num: number = 1;
  public run: number = 0;
  public trap: number = 0;
  public specialShiftX: number = 0;
  public specialShiftY: number = 0;

  // Optional properties with defaults
  public talk: CanTalk = CanTalk.NO;
  public chat: CanChat = CanChat.NO;
  public rope: CanRope = CanRope.YES;
  public prot: CanProt = CanProt.YES;
  public wing: HasWing = HasWing.NO;
  public fire: FireRes = FireRes.NO;
  public item: HasItem = HasItem.NONE;
  public special: Special = Special.NONE;
  public mental: Mental = Mental.NORMAL;

  public attackSound: PS1Sound | null = null;
  public type: EnemyType | null = null;

  public strAnimCHR: string | null = null;
  public animCHR: CHR | null = null;
  public vertical: number = 0;
  public contact: number = 0;

  constructor(name: string) {
    this.name = name;
  }

  // Builder pattern methods - direct port of Java
  public setHp(hp: number): Enemy {
    this.hp = hp;
    return this;
  }

  public setAtk(atk: number): Enemy {
    this.atk = atk;
    return this;
  }

  public setDef(def: number): Enemy {
    this.def = def;
    return this;
  }

  public setExp(exp: number): Enemy {
    this.exp = exp;
    return this;
  }

  public setMst(mst: number): Enemy {
    this.mst = mst;
    return this;
  }

  public setNum(num: number): Enemy {
    this.num = num;
    return this;
  }

  public setRun(run: number): Enemy {
    this.run = run;
    return this;
  }

  public setTrap(trap: number): Enemy {
    this.trap = trap;
    return this;
  }

  // Optional property setters
  public setItem(item: HasItem): Enemy {
    this.item = item;
    return this;
  }

  public setTalk(talk: CanTalk): Enemy {
    this.talk = talk;
    return this;
  }

  public setChat(chat: CanChat): Enemy {
    this.chat = chat;
    return this;
  }

  public setRope(rope: CanRope): Enemy {
    this.rope = rope;
    return this;
  }

  public setProt(prot: CanProt): Enemy {
    this.prot = prot;
    return this;
  }

  public setWing(wing: HasWing): Enemy {
    this.wing = wing;
    return this;
  }

  public setFire(fire: FireRes): Enemy {
    this.fire = fire;
    return this;
  }

  public setSpecial(special: Special): Enemy {
    this.special = special;
    return this;
  }

  public setMental(mental: Mental): Enemy {
    this.mental = mental;
    return this;
  }

  public setSpcpoint(x: number, y: number): Enemy {
    this.specialShiftX = x;
    this.specialShiftY = y;
    return this;
  }

  public setSound(sound: PS1Sound): Enemy {
    this.attackSound = sound;
    return this;
  }

  public setAnim(strAnim: string): Enemy {
    this.strAnimCHR = strAnim;
    return this;
  }

  public setVertical(vertical: number): Enemy {
    this.vertical = vertical;
    return this;
  }

  public setContact(contact: number): Enemy {
    this.contact = contact;
    return this;
  }

  public setType(type: EnemyType): Enemy {
    this.type = type;
    return this;
  }

  // Getters - direct port of Java
  public getSound(): PS1Sound | null {
    return this.attackSound;
  }

  public getMaxHp(): number {
    return this.hp;
  }

  public getMaxNum(): number {
    return this.num;
  }

  // Lazy load CHR
  public async getChr(scene?: Phaser.Scene, basePath: string = 'src/demos/ps'): Promise<CHR | null> {
    if (this.animCHR === null && this.strAnimCHR && scene) {
      try {
        this.animCHR = await CHR.loadChr(scene, this.strAnimCHR, basePath);
        console.log(`Enemy CHR loaded: ${this.strAnimCHR}`);
      } catch (error) {
        console.error(`Failed to load enemy CHR ${this.strAnimCHR}:`, error);
      }
    }
    return this.animCHR;
  }

  public getMental(): number {
    switch (this.mental) {
      case Mental.HIGHER:
        return this.getMaxHp() * 2;
      case Mental.LOWER:
        return Math.floor(this.getMaxHp() / 2);
      case Mental.LOWEST:
        return Math.floor(this.getMaxHp() / 4);
      default:
        return this.getMaxHp();
    }
  }

  public getName(): string {
    return PSGame.getString(this.name);
  }
}